### PR TITLE
chore: bump MSRV to 1.73

### PR DESCRIPTION
**Pull request by thomaseizinger**
## Description

With the upcoming `v0.53` release, I'd suggest we also bump the MSRV. This will allows to make use of new things like:

- `let else`
- `std::pin::pin!`
- `OnceCell`

For now, this PR doesn't make us of any of these. We can always ship them as internal improvements later and enforce certain things via lints (like `let else`).

## Notes & open questions

Should we not bump to the latest version but something before that? Do we know of any users that might be constrained by MSRV requirements from updating?

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

**Review by mxinden — APPROVED** [review]
> Should we not bump to the latest version but something before that? Do we know of any users that might be constrained by MSRV requirements from updating?

Not aware of any. I suggest bumping to the latest version.

File: libp2p/CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,7 @@
 ## 0.53.0 - unreleased
 
+- Raise MSRV to 1.73.
+  See [PR XXXX](https://github.com/libp2p/rust-libp2p/pull/XXXX).
```
Comment (thomaseizinger) [inline libp2p/CHANGELOG.md]:
```suggestion
  See [PR 4692](https://github.com/libp2p/rust-libp2p/pull/4692).
```

File: Cargo.toml [hunk 2]
```diff
@@ -135,6 +135,7 @@ clippy.used_underscore_binding = "warn"
 clippy.pedantic = "allow"
 clippy.type_complexity = "allow"
 clippy.unnecessary_wraps = "warn"
+clippy.manual_let_else = "warn"
```
Comment (thomaseizinger) [inline Cargo.toml]:
```suggestion
```

**Comment by thomaseizinger** [comment]
@mxinden I had to modify the interop-tests docker image to statically link glibc because the 1.73 image comes with a different version from what is within the chromium image.

File: interop-tests/Dockerfile.chromium [hunk 3]
```diff
@@ -14,13 +14,13 @@ FROM chef AS builder
 COPY --from=planner /app/recipe.json recipe.json
 # Build dependencies - this is the caching Docker layer!
 RUN cargo chef cook --release --package interop-tests --target wasm32-unknown-unknown --recipe-path recipe.json
-RUN cargo chef cook --release --package interop-tests --bin wasm_ping --recipe-path recipe.json
+RUN RUSTFLAGS='-C target-feature=+crt-static' cargo chef cook --release --package interop-tests --target x86_64-unknown-linux-gnu --bin wasm_ping --recipe-path recipe.json
```
Comment (mxinden) [inline interop-tests/Dockerfile.chromium:17]: Can you add a short comment why this is needed? For future us to eventually go back to the provided glibc?

**Review by mxinden — COMMENTED** [review]
Thanks for finishing this up.

**Comment by SuperFluffy** [comment]
We are using libp2p transitively through one of our deps, and this change requires us to bump our entire msrv to 1.73.

I am trying to understand the motivation behind using latest stable as msrv - from the 3 listed newer rust features I believe once cell is the newest in 1.70. `let ... else` and pin! were stabilized some time in the 1.60s. Is there a feature I am missing?

It's not a big issue for us (other than being a chore), but we are trying to be as loose as possible with msrv's.

**Comment by thomaseizinger** [comment]
Personally, one of the more exciting features in Rust that I am looking forward to is async-fn-in-traits which will stabilize in 1.75.

I'd like to bump to that with the next minor release.

In regards to the `v0.53` release, we could have probably gotten away with a more conservative bump. Frankly, not that much thought went into it. We don't have a strict MSRV policy other than considering them breaking changes and hence only bumping them in minor releases.

**Pull request merged**